controllers/containerizedworkload_controller.go

```diff
@@ -0,0 +1,156 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"time"
+
+	cpv1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/go-logr/logr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	oamv1alpha2 "github.com/oam-dev/core-resource-controller/api/v1alpha2"
+)
+
+const (
+	oamReconcileWait = 30 * time.Second
+)
+
+// Reconcile error strings.
+const (
+	errRenderWorkload  = "cannot render workload"
+	errUpdateStatus    = "cannot apply status"
+	errApplyDeployment = "cannot apply the deployment"
+	errApplyService    = "cannot apply the service"
+	errGCDeployment    = "cannot clean up stale deployments"
+	errScaleDeployment = "cannot scale the deployment"
+)
+
+// ContainerizedWorkloadReconciler reconciles a ContainerizedWorkload object
+type ContainerizedWorkloadReconciler struct {
+	client.Client
+	Log    logr.Logger
```

Comment:
I tried it but it somehow doesn't have log.Error(). I would prefer to have an Error (actually better to have error/info/debug)

Reply:
Let me try to sell you on eschewing `log.Error`. If you're not convinced we can use `logr` here instead. :)

https://dave.cheney.net/2015/11/05/lets-talk-about-logging
https://peter.bourgon.org/go-best-practices-2016/#logging-and-instrumentation

Crossplane's logging package is inspired by the above articles (dare I say best practices?). I feel that `log.Error(err, "whoops")` is about the same amount of typing as `log.Debug("whoops", "error", err)` or `log.Info("whoops", "error", err)`. I like that using `Info` or `Debug` to record errors allows us to avoid presuming that errors should always be logged at or above info level. It frequently makes sense to log errors at debug and emit a Kubernetes warning event. Often when a controller encounters an error said error pertains to a particular custom resource, and is thus interesting to the folks interacting with that resource and not to the folks actually running the controller process.